File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
    ArrayList<String> result= new ArrayList<String>();
    this.recursive(new Stack<String>(), result, 0, 0, n);
    return result;
    }

    public void recursive(Stack<String> currString, ArrayList<String> result, int open, int close, int n) {
        if(currString.size()==2*n){
            result.add(String.join("",currString));
            return;
        }
        if(open<n){
            currString.push("(");
            this.recursive(currString, result, open+1, close, n);
            currString.pop();
        }
        if(open > close){
            currString.push(")");
            this.recursive(currString, result, open, close+1, n);
            currString.pop();
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result_str_list=[]
        self.recursive(n, 0, [], result_str_list)
        return result_str_list

    def recursive(self, n, open_count, curr_list, result_str_list):
        if len(curr_list)==2*n:
            if open_count==0:
                result_str_list.append(''.join(curr_list))
            return

        if open_count<n:
            curr_list.append('(')
            self.recursive(n, open_count+1, curr_list, result_str_list)
            curr_list.pop()               
        
        if open_count>0:
            curr_list.append(')')
            self.recursive(n, open_count-1, curr_list, result_str_list)
            curr_list.pop()
            